diskqueue.go

```diff
@@ -395,6 +422,188 @@ func (d *diskQueue) readOne() ([]byte, error) {
+		// only go through files and skip directories
+		if !fileInfo.IsDir() {
+			err = fn(fileInfo)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (d *diskQueue) getAllBadFileInfo() ([]os.FileInfo, error) {
+	var badFileInfos []os.FileInfo
+
+	getAllBadFileInfo := func(fileInfo os.FileInfo) error {
+		// only accept "bad" files created by this DiskQueue object
+		if badFileNameRegexp.MatchString(fileInfo.Name()) {
+			badFileInfos = append(badFileInfos, fileInfo)
+		}
+
+		return nil
+	}
+
+	err := d.walkDiskQueueDir(getAllBadFileInfo)
+
+	return badFileInfos, err
+}
+
+// get the accurate total non-"bad" file size
+func (d *diskQueue) updateTotalDiskSpaceUsed() error {
+	d.totalDiskSpaceUsed = maxMetaDataFileSize
+
+	updateTotalDiskSpaceUsed := func(fileInfo os.FileInfo) error {
+		// only accept files created by this DiskQueue object
+		if fileNameRegexp.MatchString(fileInfo.Name()) || badFileNameRegexp.MatchString(fileInfo.Name()) {
+			d.totalDiskSpaceUsed += fileInfo.Size()
+		}
+
+		return nil
+	}
+
+	return d.walkDiskQueueDir(updateTotalDiskSpaceUsed)
+}
+
+func (d *diskQueue) freeDiskSpace(expectedBytesIncrease int64) error {
+	var err error
+	var badFileInfos []os.FileInfo
+
+	badFileInfos, err = d.getAllBadFileInfo()
+	if err != nil {
+		d.logf(ERROR, "DISKQUEUE(%s) failed to retrieve all .bad file info - %s", d.name, err)
+	}
+
+	// keep freeing up disk space until we have enough space to write this message
+	for d.totalDiskSpaceUsed+expectedBytesIncrease > d.maxBytesDiskSpace {
+		if len(badFileInfos) > 0 {
+			// check if a .bad file exists. If it does, delete that first
+			oldestBadFileInfo := badFileInfos[0]
+			badFileFilePath := path.Join(d.dataPath, oldestBadFileInfo.Name())
+
+			err = os.Remove(badFileFilePath)
+			if err != nil {
+				d.logf(ERROR, "DISKQUEUE(%s) failed to remove .bad file(%s) - %s", d.name, oldestBadFileInfo.Name(), err)
+				d.updateTotalDiskSpaceUsed()
+				return err
+			} else {
+				// recaclulate total bad files disk size to get most accurate info
+				d.totalDiskSpaceUsed -= oldestBadFileInfo.Size()
+				d.logf(INFO, "DISKQUEUE(%s) removed .bad file(%s) of size(%d bytes) to free up disk space", d.name, oldestBadFileInfo.Name(), oldestBadFileInfo.Size())
+			}
+
+			badFileInfos = badFileInfos[1:]
+		} else {
+			// delete the read file (make space)
+			readFileToDeleteNum := d.readFileNum
+			err = d.removeReadFile()
+			if err != nil {
+				d.logf(ERROR, "DISKQUEUE(%s) failed to remove file(%s) - %s", d.name, d.fileName(readFileToDeleteNum), err)
+				d.handleReadError()
+				return err
+			} else {
+				d.logf(INFO, "DISKQUEUE(%s) removed file(%s) to free up disk space", d.name, d.fileName(readFileToDeleteNum))
+			}
+			d.updateTotalDiskSpaceUsed()
+		}
+	}
+
+	return nil
+}
+
+// check if there is enough available disk space to write new data to file
+func (d *diskQueue) checkDiskSpace(totalBytes int64, reachedFileSizeLimit bool) error {
```

Comment:
I would like `func (d *diskQueue) checkDiskSpace(totalBytes int64)`.   

Add numFileMsgBytes to totalBytes before call checkDiskSpace if numFileMsgBytes should be added to it.

Reply:
LOL very good catch! I cannot believe I missed this Q.Q

Reply:
Sure.